joinfaces-autoconfigure/src/main/java/org/joinfaces/autoconfigure/rewrite/SpringBootAnnotationConfigProvider.java

```diff
@@ -0,0 +1,142 @@
+ * <p>
+ * Inspired by the {@code AnnotationConfigProvider}: Finds class' names and handles the rule evaluation via
+ * {@link ClassVisitorImpl}.
+ * <p>
+ * Note: This scanner is disabled by default to stay compliant with the original behaviour and because the classpath
+ * scanning is very time costly if the packages are not narrowed down.
+ *
+ * @author Patrick Mentz
+ * @see AnnotationConfigProvider
+ * @see org.ocpsoft.rewrite.annotation.scan.WebClassesFinder
+ */
+@Slf4j
+public class SpringBootAnnotationConfigProvider extends HttpConfigurationProvider {
+
+	public static final String SCAN_CLASSPATH = "org.joinfaces.rewrite.annotation.SCAN_CLASSPATH";
+
+	@Override
+	public Configuration getConfiguration(final ServletContext servletContext) {
+
+		if (!Boolean.valueOf(servletContext.getInitParameter(SCAN_CLASSPATH))) {
+			log.debug("Annotation scanning in classpath is disabled!");
+			return null;
+		}
+
+		final String packageFilters = Optional.ofNullable(
+				servletContext.getInitParameter(AnnotationConfigProvider.CONFIG_BASE_PACKAGES)).orElse("").trim();
+		// rewrite allows disable annotation scanning globally
+		if (packageFilters.equalsIgnoreCase("none")) {
+			log.debug("Annotation scanning is globally disabled!");
+			return null;
+		}
+		else if (packageFilters.isEmpty()) {
+			log.warn("No base package defined, annotation scanning may be extremely slow");
+		}
+
+		// Generate a list of all relevant annotations
+		final Set<Class<? extends Annotation>> ruleAnnotations = new LinkedHashSet<>();
+		final List<AnnotationHandler<Annotation>> annotationHandlers = new ArrayList<>();
+		@SuppressWarnings("unchecked")
+		final Iterator<AnnotationHandler<Annotation>> handlerIterator = ServiceLoader.load(
+				AnnotationHandler.class).iterator();
+		while (handlerIterator.hasNext()) {
+			final AnnotationHandler<Annotation> handler = handlerIterator.next();
+			annotationHandlers.add(handler);
+			ruleAnnotations.add(handler.handles());
+		}
+
+		final ClassVisitorImpl ruleBuilderVisitor = new ClassVisitorImpl(annotationHandlers, servletContext);
+
+		scanClasses(packageFilters.split("\\s*,\\s*"), ruleAnnotations, ruleBuilderVisitor);
+		return ruleBuilderVisitor;
+	}
+
+	@Override
+	public int priority() {
+		return 100;
+	}
+
+	/**
+	 * Scans classes in the given package for one of the annotations given and calls the given visitor on them.
+	 *
+	 * @param basePackages the packages to scan
+	 * @param supportedAnnotations the annotations to search for
+	 * @param visitor the visitor to trigger with matching classes
+	 */
+	private void scanClasses(final String[] basePackages,
+			final Set<Class<? extends Annotation>> supportedAnnotations,
+			final ClassVisitor visitor) {
+		final ClassPathScanningCandidateComponentProvider scanner = new ClassPathScanningCandidateComponentProvider(false);
```

Comment:
Have you tried to use Classgraph instead? Classgraph is more flexible than `ClassPathScanningCandidateComponentProvider` and also faster (especially when scanning the whole classpath).

We might even be able to perfom this scan at buildtime (as we already do for MyFaces and `ServletContainerInitializer`s)

Reply:
I did go for the `ClassPathScanningCandidateComponentProvider` as it was the common approach in a Spring environment for me, it also provides support for Spring meta annotations. But I don't think meta annotations make any sense when it comes to Rewrite, so I will do some tests regarding performance. If Classgraph seems to be the more reasonable approach, my question would be if this should not be a PR to Rewrite itself, to change their overall approach to find annotations.

Regarding the scan during build time, this sounds like a nice idea, but I would need further support on that one.

Reply:
Ok, if a base-package is defined, the scanning with Classgraph is a little bit slower (we are talking about 60ms vs 30ms), but without base-package defined, which triggers an overall scan, Classgraph is much faster (600ms vs 10600ms). So I switched to Classgraph, thank you very much for that hint, that was quite interesting.

Reply:
The best approach would be if this todo would be resolved: https://github.com/ocpsoft/rewrite/blob/eded0bad0101d119a3a1e67ed3f97a3bc9a8222e/annotations-impl/src/main/java/org/ocpsoft/rewrite/annotation/config/AnnotationConfigProvider.java#L130-L131

If we could inject our own `ClassFinder` implementation there, our solution would have a smaller footprint. On the other hand, the bytecode and package filtering done by Rewrite could be done by Classgraph too.